### PR TITLE
Add event reporter

### LINKED ISSUE #477: Add error reporting in topology builder and provider
In the (upcoming) files [`topology/builder.go`](https://github.com/jspdown/maesh/blob/acl-provider/pkg/topology/builder.go) and [`provider/provider.go`](https://github.com/jspdown/maesh/blob/acl-provider/pkg/provider/provider.go), we should replace the logger by some kind of error reporting mechanism that not only logs the errors but also aggregates them, in order to have a list of the issues present in the dynamic configuration.

This could be used later on to add an entrypoint to the API to list the configuration errors with precise context.

### PR DESCRIPTION
**Pull request by LandryBe**
### What does this PR do?

This PR introduces a new system to report errors enriched with some context.

In a future PR we will be able to leverage this new system and add an endpoint to the API to list those errors.

Fixes #477

### CODE REVIEW

File: pkg/topology/builder.go [hunk 1]
```diff
@@ -45,16 +45,20 @@ func (b *Builder) Build(ignoredResources mk8s.IgnoreWrapper) (*Topology, error)
...
 		}
 	}
 
 	// Populate services with traffic-split definitions.
-	for key, ts := range res.TrafficSplits {
+	for _, ts := range res.TrafficSplits {
 		if err := b.evaluateTrafficSplit(topology, ts); err != nil {
-			b.Logger.Errorf("Unable to evaluate TrafficSplit %q: %v", key, err)
+			b.Logger.
+				ForSubject(ts.Namespace, "TrafficTarget", ts.Name).
```
Comment (jspdown) [inline pkg/topology/builder.go:60]:
```suggestion
				ForSubject(ts.Namespace, "TrafficSplit", ts.Name).
```

File: pkg/topology/builder.go [hunk 2]
```diff
@@ -237,15 +241,20 @@ func (b *Builder) populateTrafficSplitsAuthorizedIncomingTraffic(topology *Topol
 		for _, tsKey := range svc.TrafficSplits {
 			ts, ok := topology.TrafficSplits[tsKey]
 			if !ok {
-				b.Logger.Errorf("Unable to find TrafficSplit %q", tsKey)
+				b.Logger.
+					ForSubject(ts.Namespace, "TrafficTarget", ts.Name).
```
Comment (jspdown) [inline pkg/topology/builder.go:245]:
```suggestion
					ForSubject(ts.Namespace, "TrafficSplit", ts.Name).
```

File: pkg/topology/builder.go [hunk 3]
```diff
@@ -237,15 +241,20 @@ func (b *Builder) populateTrafficSplitsAuthorizedIncomingTraffic(topology *Topol
...
 				continue
 			}
 
 			pods, err := b.getIncomingPodsForTrafficSplit(topology, ts, map[Key]struct{}{})
 			if err != nil {
 				loopDetected[svc] = append(loopDetected[svc], tsKey)
 
-				b.Logger.Errorf("Unable to get incoming pods for TrafficSplit %q: %v", tsKey, err)
+				b.Logger.
+					ForSubject(ts.Namespace, "TrafficTarget", ts.Name).
```
Comment (jspdown) [inline pkg/topology/builder.go:256]:
```suggestion
					ForSubject(ts.Namespace, "TrafficSplit", ts.Name).
```

File: pkg/provider/provider.go [hunk 4]
```diff
@@ -289,7 +289,7 @@ func (p *Provider) buildServiceAndRoutersForTrafficSplit(t *topology.Topology, c
 
 	logger := p.logger.
 		WithMetadata(tsSvc.Annotations).
```
Comment (jspdown) [inline pkg/provider/provider.go:291]: Since you updated the Subject to TrafficSplit, metadata should be updated too, no?

**Comment by LandryBe** [comment]
As described in #522, we are closing this PR and will open another one with errors propagated in the topology itself.

**Pull request closed without merging**